lib/datadog/appsec/contrib/rack/gateway/watcher.rb

```diff
@@ -110,6 +111,25 @@ def watch_request_body(gateway = Instrumentation.gateway)
                   stack.call(gateway_request.request)
                 end
               end
+
+              # NOTE: In the current state we unable to substibe twice to the same
+              #       event within the same group. Ideally this code should live
+              #       somewhere closer to identity related monitor.
+              # WARNING: The Gateway is a subject of refactoring
+              def watch_request_finish(gateway = Instrumentation.gateway)
+                gateway.watch('rack.request.finish', :appsec) do |stack, gateway_request|
+                  context = gateway_request.env[AppSec::Ext::CONTEXT_KEY]
+                  next stack.call(gateway_request.request) if context.span.nil? || !gateway.pushed?('identity.set_user')
+
+                  gateway_request.headers.each do |name, value|
+                    next unless Ext::IDENTITY_COLLECTABLE_REQUEST_HEADERS.include?(name)
+
+                    context.span["http.request.headers.#{name}"] = value
+                  end
+
+                  stack.call(gateway_request.request)
```

Comment:
Do we want to call the next item in the stack with `gateway_request.request`, or just `gateway_request`? I would assume that if we have 2 watchers for the same event, they would get the same argument?

Reply:
It's the same as in [the method above](https://github.com/DataDog/dd-trace-rb/pull/4468/files#diff-77759fd9cacb661757fbe67537b41c78174137c8cda751f73d0be8746602ddb6R111). I know that it looks subpar, but it's about to change soon (I hope)